### ISSUE TITLE
M2Crypto workaround

### ISSUE THREAD

**Issue by bari12** [cern-fts/fts-rest#1]
Hi,
The way the M2Crpyto workaround is done in the `setup.py` is quite problematic, since it bypasses any dependency specification by pip.

We install the fts3-rest-api among other dependencies including M2Crypto specified in a dependency file. This is installed in a loop by pip. But since this workaround calls pip again (from inside the pip loop) it kills any dependency selection done for M2Crypto in the outer installation loop.

**Comment by andrea-manzi** [comment]
thanks for reporting 

i'm going to check how we can better implement this workaround for m2crypto

**Comment by andrea-manzi** [comment]
any issue with the pycurl workaround? as it also runs pip again

**Comment by bari12** [comment]
Pycurl workaround is also not good, but doesn't hit us that hard since we are on CC7 and don't specify pycurl as dependency directly. 
But for M2Crypto it was really nasty, since the actual pip loop was installing a different version than the pip call within the fts3-rest setup.py. This came only up since M2Crpyo recently released a new version. But this way we ended up with a mixture of files from two different versions within the M2Crpyto directory, which crashed everything.

**Comment by andrea-manzi** [comment]
Hi Martin
i investigated a bit and the workaround we are applying for the installation of m2crypto it's only needed on Sl6..so although is not the best solution we can run it only for that OS and this would solve your issue.
If i apply the change in a branch would you be able to test it?
thanks
Andrea

**Comment by andrea-manzi** [comment]
i just did the change and it seems ok C7

```

root@fts-dev-andrea ~]# virtualenv fts-rest-test
 New python executable in /root/fts-rest-test/bin/python
 Installing setuptools, pip, wheel...
 done.
[root@fts-dev-andrea ~]# source fts-rest-test/bin/activate
 (fts-rest-test) [root@fts-dev-andrea ~]# 
 (fts-rest-test) [root@fts-dev-andrea ~]# 
 (fts-rest-test) [root@fts-dev-andrea ~]# pip install "git+https://gitlab.cern.ch/fts/fts-rest.git@FTS-1435"
 Collecting git+https://gitlab.cern.ch/fts/fts-rest.git@FTS-1435
 Cloning https://gitlab.cern.ch/fts/fts-rest.git (to FTS-1435) to /tmp/pip-MhVP9X-build
 Collecting M2Crypto>=0.16 (from fts3-rest==3.9.1)
 Downloading https://files.pythonhosted.org/packages/74/18/3beedd4ac48b52d1a4d12f2a8c5cf0ae342ce974859fba838cbbc1580249/M2Crypto-0.35.2.tar.gz (1.1MB)
 100% |████████████████████████████████| 1.1MB 799kB/s 
 Collecting pycurl==7.19.0 (from fts3-rest==3.9.1)
 Collecting requests (from fts3-rest==3.9.1)
 Downloading https://files.pythonhosted.org/packages/51/bd/23c926cd341ea6b7dd0b2a00aba99ae0f828be89d72b2190f27c11d4b7fb/requests-2.22.0-py2.py3-none-any.whl (57kB)
 100% |████████████████████████████████| 61kB 8.1MB/s 
 Collecting typing (from M2Crypto>=0.16->fts3-rest==3.9.1)
 Downloading https://files.pythonhosted.org/packages/79/be/e43ccd1317f483e14de507787322c74092a688707d02fa5f16562270eb63/typing-3.7.4-py2-none-any.whl
 Collecting chardet<3.1.0,>=3.0.2 (from requests->fts3-rest==3.9.1)
 Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
 Collecting idna<2.9,>=2.5 (from requests->fts3-rest==3.9.1)
 Downloading https://files.pythonhosted.org/packages/14/2c/cd551d81dbe15200be1cf41cd03869a46fe7226e7450af7a6545bfc474c9/idna-2.8-py2.py3-none-any.whl (58kB)
 100% |████████████████████████████████| 61kB 8.7MB/s 
 Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 (from requests->fts3-rest==3.9.1)
 Downloading https://files.pythonhosted.org/packages/e6/60/247f23a7121ae632d62811ba7f273d0e58972d75e58a94d329d51550a47d/urllib3-1.25.3-py2.py3-none-any.whl (150kB)
 100% |████████████████████████████████| 153kB 6.1MB/s 
 Collecting certifi>=2017.4.17 (from requests->fts3-rest==3.9.1)
 Downloading https://files.pythonhosted.org/packages/69/1b/b853c7a9d4f6a6d00749e94eb6f3a041e342a885b87340b79c1ef73e3a78/certifi-2019.6.16-py2.py3-none-any.whl (157kB)
 100% |████████████████████████████████| 163kB 6.1MB/s 
 Building wheels for collected packages: M2Crypto
 Running setup.py bdist_wheel for M2Crypto ... done
 Stored in directory: /root/.cache/pip/wheels/6d/b3/6e/4c057fb6aafc93b1a8d5607d46da439260ae21dbca3e9ceba8
 Successfully built M2Crypto
 Installing collected packages: typing, M2Crypto, pycurl, chardet, idna, urllib3, certifi, requests, fts3-rest
 Running setup.py install for fts3-rest ... done
 Successfully installed M2Crypto-0.35.2 certifi-2019.6.16 chardet-3.0.4 fts3-rest-3.9.1 idna-2.8 pycurl-7.19.0 requests-2.22.0 typing-3.7.4 urllib3-1.25.3
 You are using pip version 9.0.1, however version 19.1.1 is available.
 You should consider upgrading via the 'pip install --upgrade pip' command.
 (fts-rest-test) [root@fts-dev-andrea ~]#
(fts-rest-test) [root@fts-dev-andrea ~]# voms-proxy-init -voms dteam
 Contacting voms2.hellasgrid.gr:15004 [/C=GR/O=HellasGrid/OU=hellasgrid.gr/CN=voms2.hellasgrid.gr] "dteam"...
 Remote VOMS server contacted succesfully.
Created proxy in /tmp/x509up_u0.
Your proxy is valid until Fri Jun 21 00:14:44 CEST 2019
 
(fts-rest-test) [root@fts-dev-andrea ~]# which fts-rest-whoami
 
(fts-rest-test) [root@fts-dev-andrea ~]# fts-rest-whoami 
 User DN: /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=amanzi/CN=683749/CN=Andrea Manzi
 VO: dteam
 VO id: 6b10f4e4-8fdc-5555-baa2-7d4850d4f406
 Delegation id: bfd3b0c180c099ea
 Base id: 01874efb-4735-4595-bc9c-591aef8240c9
 (fts-rest-test) [root@fts-dev-andrea ~]# which fts-rest-whoami 
 /root/fts-rest-test/bin/fts-rest-whoami
 
(fts-rest-test) [root@fts-dev-andrea ~]# fts-rest-delegate 
 Delegation id: bfd3b0c180c099ea
 
 

 

```

**Comment by bari12** [comment]
Hi,
Yes this would work for us since we don't use SE6 anymore.
Cheers,
Martin

**Comment by andrea-manzi** [comment]
ok i pushed the changes on our develop branch, to be included on v 3.9.1
